parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +607,342 @@ protected GroupType build(String name) {
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> optionalList() {
+      return list(Type.Repetition.OPTIONAL);
+    }
+  }
+
+  /**
+   * A builder for {@link GroupType} objects.
+   *
+   * @param <P> The type that this builder will return from
+   *          {@link #named(String)} when the type is built.
+   */
+  public static class GroupBuilder<P> extends BaseGroupBuilder<P, GroupBuilder<P>> {
+
+    private GroupBuilder(P parent) {
+      super(parent);
+    }
+
+    private GroupBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected GroupBuilder<P> self() {
+      return this;
+    }
+  }
+
+  public static class MapBuilder<P> extends Builder<MapBuilder<P>, P> {
+    private static final Type STRING_KEY = Types
+        .required(PrimitiveTypeName.BINARY).as(OriginalType.UTF8).named("key");
+
+    public static class MapKeyBuilder<P, Q> extends PrimitiveBuilder<P> {
+      private final P parent;
+      private final Q grandParent;
+
+      public MapKeyBuilder(P parent, Q grandParent, PrimitiveTypeName type) {
+        super(parent, type);
+        this.parent = parent;
+        this.grandParent = grandParent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      @Override
+      public MapKeyBuilder<P, Q> as(OriginalType orig) {
+        return (MapKeyBuilder<P, Q>) super.as(orig);
+      }
+
+      public MapValueBuilder<P, Q> value(PrimitiveTypeName type, Type.Repetition repetition) {
+        ((MapBuilder)parent).setKeyType(build("key"));
+        if (grandParent != null) {
+          return (MapValueBuilder<P, Q>) new MapValueBuilder<P, Q>(parent, grandParent,
+              type).repetition(repetition);
+        } else {
+          return (MapValueBuilder<P, Q>) new MapValueBuilder<P, Q>(parent,
+              ((MapBuilder) parent).returnType, type).repetition(repetition);
+        }
+      }
+
+      public MapValueBuilder<P, Q> requiredValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P, Q> optionalValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.OPTIONAL);
+      }
+
+      public MapGroupValueBuilder<P, Q> groupValue(Type.Repetition repetition) {
+        ((MapBuilder)parent).setKeyType(build("key"));
+        return (MapGroupValueBuilder<P, Q>) new MapGroupValueBuilder<P, Q>(parent,
+            grandParent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P, Q> requiredGroupValue() {
+        return groupValue(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P, Q> optionalGroupValue() {
+        return groupValue(Type.Repetition.OPTIONAL);
+      }
+
+      @Override
+      public P named(String name) {
+        throw new UnsupportedOperationException("Can not assign name on MapKeyBuilder");
+      }
+    }
+
+    public static class MapValueBuilder<P, Q> extends PrimitiveBuilder<Q> {
```

Comment:
Inheriting from `PrimitiveBuilder` directly is problematic because methods return `PrimitiveBuilder` and not the `MapValueBuilder`. That means you can't call `as` and then the `value` methods. I think you'll need a `BasePrimitiveBuilder` like you have `BaseGroupBuilder` that sets up the `self()` method to properly return the right builder type.